File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/Solution_Task4_ErrorHandling.java
package com.jraska.rx.codelab.solution;

import com.jraska.rx.codelab.RxLogging;
import com.jraska.rx.codelab.http.HttpBinApi;
import com.jraska.rx.codelab.http.HttpModule;
import io.reactivex.internal.functions.Functions;
import okhttp3.MediaType;
import okhttp3.ResponseBody;
import org.junit.Before;
import org.junit.Test;

public class Solution_Task4_ErrorHandling {
  private HttpBinApi httpBinApi = HttpModule.httpBinApi();

  @Before
  public void before() {
    RxLogging.INSTANCE.enableObservableSubscribeLogging();
  }

  @Test
  public void printErrorMessage() {
    httpBinApi.failingGet()
      .subscribe(System.out::println, System.err::println);
  }

  @Test
  public void onErrorReturnItem_emitCustomItemOnError() {
    httpBinApi.failingGet()
      .onErrorReturnItem(syntheticBody())
      .subscribe(System.out::println);
  }

  @Test
  public void onErrorResumeNext_subscribeToExtraObservableOnError() {
    httpBinApi.failingGet()
      .onErrorResumeNext(httpBinApi.backupGet())
      .subscribe(System.out::println, Functions.emptyConsumer());
  }

  @Test
  public void retry_retryOnError() {
    httpBinApi.flakyGet()
      .retry()
      .subscribe(System.out::println);
  }

  static ResponseBody syntheticBody() {
    return ResponseBody.create(MediaType.get("application/json"), "{}");
  }
}

File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/SolutionTask4ErrorHandling.kt
package com.jraska.rx.codelab.solution

import com.jraska.rx.codelab.RxLogging
import com.jraska.rx.codelab.http.HttpModule
import io.reactivex.functions.Consumer
import io.reactivex.internal.functions.Functions
import okhttp3.MediaType
import okhttp3.ResponseBody
import org.junit.Before
import org.junit.Test

class SolutionTask4ErrorHandling {
  private val httpBinApi = HttpModule.httpBinApi()

  @Before
  fun before() {
    RxLogging.enableObservableSubscribeLogging()
  }

  @Test
  fun printErrorMessage() {
    httpBinApi.failingGet()
      .subscribe(System.out::println, System.err::println)
  }

  @Test
  fun onErrorReturnItem_emitCustomItemOnError() {
    httpBinApi.failingGet()
      .onErrorReturnItem(syntheticBody())
      .subscribe { println(it) }
  }

  @Test
  fun onErrorResumeNext_subscribeToExtraObservableOnError() {
    httpBinApi.failingGet()
      .onErrorResumeNext(httpBinApi.backupGet())
      .subscribe(Consumer { println(it) }, Functions.emptyConsumer())
  }

  @Test
  fun retry_retryOnError() {
    httpBinApi.flakyGet()
      .retry()
      .subscribe { println(it) }
  }

  companion object {
    fun syntheticBody(): ResponseBody {
      return ResponseBody.create(MediaType.get("application/json"), "{}")
    }
  }
}
